### ISSUE TITLE
Unique IDs for variants (variant server)

### ISSUE THREAD

**Issue by BoPeng** [VariantAssociationTools/VAT#2]
One of the major problems that prevents us from distributing data in separate projects (locations, servers) are that there is no unified index for variants. For example, we will have to sort variants and compare variants from different projects and reassign IDs to them when we merge projects, which is painfully slow. 

There are seemingly two solutions:

1. Use `chrom`, `pos`, `ref`, `alt` to store and compare variants whenever variants are stored. This is likely to be slow but it might be ok if the databases are properly indexed (with four fields). It is unclear on how to handle the same variant across different versions of reference genome, and how to handle complex genetic variants.

2. Assign an unique ID to all variants and use this master table for all projects and annotations. The problem is that such a database can be quite complicated, considering the same variant for different reference genome, and how to handle unknown variants. The IDs can be numeric or string and I am not sure there is any performance difference. We should perhaps provide a server with an API so that vat can submit new variants to the server and obtain unique IDs.

**Comment by gaow** [comment]
A recent work from (the) Rong Chen's lab seems relevant:

http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-015-0865-9

**Comment by BoPeng** [comment]
Yes. This is what I have in mind, although unfortunately it does not support for example mouse genome, and without reading the paper, I am not sure if it supports variant normalization and return/query of variants in different coordinates.

**Comment by BoPeng** [comment]
Reference Variant Store provides a lot of annotation information but I am not sure how we can make use of it. The source code is certainly not available and the REST service is quite limited.

My biggest concern is with its 64 byte KEY, which is long and make it difficult store and transfer variant information (between our DA and DP). Although it has the advantage of sorting and reverse variant decoding, I cannot seem to justify its use in VAT. It also suffers from problems such as failure to encode other types of variants, and reference genome dependency (hg37, 38 etc) and a change of reference genome might invalidate all samples and annotations.

My proposal of variant KEY would be a combination of encoding and random ID. That is to say, we use an integer of 64 bit to store variant KEY but encode partial variant information in the KEY. For example, with bytes encoding

1. species (?)
2. type of variant (SNP etc)
3. chromosome <-  1-23 etc and then manifold, special coding for cross-chromosome variant
4. band <- something like q24 in 3q24, special coding for cross-band variant
5. randomly assigned variant ID

With this coding mechanism, we can easily divide variants by type and rough location and we only need to search a relatively short region for variants (perhaps in subtables created on-the-fly). It is also possible for us to provide sorted variants by reference genome (e.g. a query to get all variants belong to certain region).

A huge burden for us would to make sure variants imported in different reference genomes are mapped to the same variants, but we can start from hg37 or 38 and design a good algorithm to convert coordinates.
